File: LEM_Listas/Listas_Enlazadas.py
class Nodo:
    def __init__(self, data):
        self.data = data
        self.next = None

class ListaEnlazada:
    def __init__(self):
        self.head = None

    def insertar_al_principio(self, data):
        nuevo = Nodo(data)
        nuevo.next = self.head
        self.head = nuevo

    def insertar_al_final(self, data):
        nuevo = Nodo(data)
        if not self.head:
            self.head = nuevo
            return
        ultimo = self.head
        while ultimo.next:
            ultimo = ultimo.next
        ultimo.next = nuevo

    def imprimir_lista(self):
        actual = self.head
        output = ""
        while actual:
            output += str(actual.data) + " -> "
            actual = actual.next
        output += "None"
        print(output)

    def buscar(self, data):
        actual = self.head
        while actual:
            if actual.data == data:
                return True
            actual = actual.next
        return False

    def eliminar(self, data):
        actual = self.head
        previo = None

        if actual and actual.data == data:
            self.head = actual.next
            return

        while actual and actual.data != data:
            previo = actual
            actual = actual.next

        if not actual:
            return

        previo.next = actual.next

if __name__ == "__main__":
    lista = ListaEnlazada()
    lista.insertar_al_final(10)
    lista.insertar_al_final(20)
    lista.insertar_al_final(30)
    print("Lista inicial:", end=" ")
    lista.imprimir_lista()

    lista.insertar_al_principio(5)
    print("Lista después de insertar al principio:", end=" ")
    lista.imprimir_lista()

    print("¿Esta el 20?", lista.buscar(20))
    print("¿Esta el 99?", lista.buscar(99))

    lista.eliminar(20)
    print("Lista después de eliminar 20:", end=" ")
    lista.imprimir_lista()

    lista.eliminar(5)
    print("Lista después de eliminar 5:", end=" ")
    lista.imprimir_lista()

File: LEM_Listas/Listas_Enlazadas.java
public class Listas_Enlazadas {

    static class Nodo {
        int data;
        Nodo next;

        public Nodo(int data) {
            this.data = data;
            this.next = null;
        }
    }

    static class ListaEnlazada {
        private Nodo head;

        public ListaEnlazada() {
            head = null;
        }

        public void insertar_al_principio(int data) {
            Nodo nuevo = new Nodo(data);
            nuevo.next = head;
            head = nuevo;
        }

        public void insertar_al_final(int data) {
            Nodo nuevo = new Nodo(data);
            if (head == null) {
                head = nuevo;
                return;
            }
            Nodo ultimo = head;
            while (ultimo.next != null) {
                ultimo = ultimo.next;
            }
            ultimo.next = nuevo;
        }

        public void imprimir_lista() {
            Nodo actual = head;
            while (actual != null) {
                System.out.print(actual.data + " -> ");
                actual = actual.next;
            }
            System.out.println("None");
        }

        public boolean buscar(int data) {
            Nodo actual = head;
            while (actual != null) {
                if (actual.data == data) return true;
                actual = actual.next;
            }
            return false;
        }

        public void eliminar(int data) {
            Nodo actual = head;
            Nodo previo = null;

            if (actual != null && actual.data == data) {
                head = actual.next;
                return;
            }

            while (actual != null && actual.data != data) {
                previo = actual;
                actual = actual.next;
            }

            if (actual == null) return;

            previo.next = actual.next;
        }
    }

    public static void main(String[] args) {
        ListaEnlazada lista = new ListaEnlazada();
        lista.insertar_al_final(10);
        lista.insertar_al_final(20);
        lista.insertar_al_final(30);
        System.out.print("Lista inicial: ");
        lista.imprimir_lista();

        lista.insertar_al_principio(5);
        System.out.print("Lista después de insertar al principio: ");
        lista.imprimir_lista();

        System.out.println("¿Esta el 20? " + lista.buscar(20));
        System.out.println("¿Esta el 99? " + lista.buscar(99));

        lista.eliminar(20);
        System.out.print("Lista después de eliminar 20: ");
        lista.imprimir_lista();

        lista.eliminar(5);
        System.out.print("Lista después de eliminar 5: ");
        lista.imprimir_lista();
    }
}